framework/test/Volo.Abp.EventBus.Tests/Volo/Abp/EventBus/EventBusTestModule.cs

```diff
@@ -1,8 +1,25 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Volo.Abp.EventBus.Distributed;
 using Volo.Abp.Modularity;
 
 namespace Volo.Abp.EventBus;
 
 [DependsOn(typeof(AbpEventBusModule))]
 public class EventBusTestModule : AbpModule
 {
+    public override void ConfigureServices(ServiceConfigurationContext context)
+    {
+        context.Services.TryAddTransient(typeof(MyGenericDistributedEventHandler<>));
+    }
+
+    public override void OnApplicationInitialization(ApplicationInitializationContext context)
+    {
+        var distributedEventBus = context.ServiceProvider.GetRequiredService<IDistributedEventBus>();
+        var scopeFactory = context.ServiceProvider.GetRequiredService<IServiceScopeFactory>();
+
+        var eventType = typeof(MySimpleEventData);
+        var myHandlerType = typeof(MyGenericDistributedEventHandler<>).MakeGenericType(eventType);
+        distributedEventBus.Subscribe(eventType, new IocEventHandlerFactory(scopeFactory, myHandlerType));
```

Comment:
In actual, the eventType is from an options:
```CSharp
var myOptions = context.ServiceProvider.GetRequiredService<IOptions<MyOption>>().Value;
foreach (var eventType in myOptions.Events)
{
    var myHandlerType = typeof(MyGenericDistributedEventHandler<>).MakeGenericType(eventType);
    distributedEventBus.Subscribe(eventType, new IocEventHandlerFactory(scopeFactory, myHandlerType));
}
```

Reply:
You can try this:

```cs
[DisableConventionalRegistration]
class MyGenericDistributedEventHandler
```

Reply:
No, `MyGenericDistributedEventHandler` is a generic type, it wouldn't be conventional registered with Dependency Injection.

And I actually need to register it with Dependency Injection. I have called `context.Services.AddTransient(typeof(MyGenericDistributedEventHandler<>))` manually.

Reply:
Did you try that?

Reply:
I think it's totally not related to conventional registration.
I need to register this type, and I've already manually done it.
There is an `OnRegister` hook on `AbpEventModule` that is the key point.
It will discover all the event handlers and add them to the options.
But I think the generic definition type should not be added.

Reply:
It's not work. `OnRegistered` callback still be called.
And I can't find any code about "OnRegister will not be called if a type has DisableConventionalRegistration". Could you show me that?

Reply:
Anyway, I think the generic definition type still should be ignored although `DisableConventionalRegistration` can work as you expected.

Reply:
https://github.com/abpframework/abp/blob/dev/framework/test/Volo.Abp.Core.Tests/Microsoft/Extensions/DependencyInjection/AbpConventionalDependencyInjectionExtensions_Tests.cs#L160

Reply:
https://github.com/abpframework/abp/pull/17465

我快要急死了，你看下你发的东西啊，那个测试是判断 ServiceCollection 的，和 OnRegistered 毫无关系。

DisableConventionalRegistration 的作用是禁止 ITransientDependency, ISingletonDependency 这些接口的自动注入。

但是这些接口本来就不对泛型生效，你不禁用它也不可能自动注入。更何况我就没继承这些接口。

但是实际上，我是需要注入这个泛型类的，所以我在ConfigureServices里写了 `context.Services.TryAddTransient(typeof(MyGenericDistributedEventHandler<>));`

然后他就会调用 `OnRegistered` 了！

Reply:
@maliming 所以理解了问题出在哪了吗？Abp把一个不带参数的泛型定义加入到Handlers里头是没有意义的，而且还引发了KafkaEventBus的异常。即使你可以找到其它途径可以让我绕开这个问题，但是根本原因还是在于Abp不应该把一个泛型定义加入到Handlers，这个问题应该被修复。

事实上这个问题我已经绕开了，`Configure<AbpDistributedEventBusOptions>`手动把泛型Handler删掉，或者把这个Handler改成非泛型的，用反射来访问对象，都是可行的解决方案。

Reply:
如何在framework的单元测试中复现问题?

Reply:
在一个有KafkaDistributedEventBus的项目里，按照`EventBusTestModule`配置就可以复现。
用RabbitMQ、Dapr之类的应该也同样可以复现。

`SubscribeHandlers(ITypeList<IEventHandler> handlers)` 时 `handlers` 包含泛型定义：
https://github.com/abpframework/abp/blob/7cd3f7eacf974a288789e7e9becad204181c7b26/framework/src/Volo.Abp.EventBus/Volo/Abp/EventBus/EventBusBase.cs#L176-L195

`Subscribe(Type eventType, IEventHandlerFactory factory)` 时 `eventType` 就会是 `TEvent`：
https://github.com/abpframework/abp/blob/7cd3f7eacf974a288789e7e9becad204181c7b26/framework/src/Volo.Abp.EventBus.Kafka/Volo/Abp/EventBus/Kafka/KafkaDistributedEventBus.cs#L96-L108

然后297行 `EventNameAttribute.GetNameOrDefault(Type eventType)` 就会得到 `null`，298行 `EventTypes.GetOrAdd(eventName, eventType);` 抛出异常：
https://github.com/abpframework/abp/blob/7cd3f7eacf974a288789e7e9becad204181c7b26/framework/src/Volo.Abp.EventBus.Kafka/Volo/Abp/EventBus/Kafka/KafkaDistributedEventBus.cs#L291-L302